File: frontend/src/config/branding.js
export const DEFAULT_BRANDING = {
  code: 'default',
  site_name: { zh_cn: 'CBCTF', en: 'CBCTF' },
  admin_name: { zh_cn: 'DASHBOARD', en: 'DASHBOARD' },
  browser_title: { zh_cn: 'CBCTF', en: 'CBCTF' },
  browser_description: {
    zh_cn: 'DEEP DIVE CTF competition platform',
    en: 'DEEP DIVE CTF competition platform',
  },
  footer_copyright: { zh_cn: '© 2026 CBCTF', en: '© 2026 CBCTF' },
  footer_icp_number: '陕ICP备2025076339号-1',
  footer_icp_link: 'https://beian.miit.gov.cn/',
  footer_contact_email: 'support@0rays.club',
  footer_github_url: 'https://github.com/0RAYS/CBCTF',
  home_logo: '/platform/logo.png',
  home_logo_alt: { zh_cn: 'CBCTF 首页 Logo', en: 'CBCTF home logo' },
  home: {
    hero: {
      title_prefix: { zh_cn: 'Dive Deep into the', en: 'Dive Deep into the' },
      title_highlight: { zh_cn: 'Cyber Security', en: 'Cyber Security' },
      title_suffix: { zh_cn: 'Challenge', en: 'Challenge' },
      subtitle: {
        zh_cn:
          'Join the elite community of hackers, compete in real-world challenges, and master the art of cybersecurity through hands-on experience.',
        en: 'Join the elite community of hackers, compete in real-world challenges, and master the art of cybersecurity through hands-on experience.',
      },
      primary_action: { zh_cn: 'START HACKING', en: 'START HACKING' },
      secondary_action: { zh_cn: 'LEARN MORE', en: 'LEARN MORE' },
    },
    challenge_types: {
      title_prefix: { zh_cn: 'Master All Aspects of', en: 'Master All Aspects of' },
      title_highlight: { zh_cn: 'Cyber Security', en: 'Cyber Security' },
      subtitle: {
        zh_cn: 'From web exploitation to binary analysis, our challenges cover every major domain of cybersecurity',
        en: 'From web exploitation to binary analysis, our challenges cover every major domain of cybersecurity',
      },
    },
    upcoming: {
      title_prefix: { zh_cn: 'Upcoming', en: 'Upcoming' },
      title_highlight: { zh_cn: 'Competitions', en: 'Competitions' },
      subtitle: {
        zh_cn: 'Register now for our upcoming CTF events and compete with the best',
        en: 'Register now for our upcoming CTF events and compete with the best',
      },
      action: { zh_cn: 'VIEW ALL CONTESTS', en: 'VIEW ALL CONTESTS' },
    },
    leaderboard: {
      title_prefix: { zh_cn: 'Top', en: 'Top' },
      title_highlight: { zh_cn: 'Performers', en: 'Performers' },
      subtitle: {
        zh_cn: 'Meet the elite teams leading our global leaderboard',
        en: 'Meet the elite teams leading our global leaderboard',
      },
      action: { zh_cn: 'VIEW SCOREBOAR', en: 'VIEW SCOREBOARD' },
    },
  },
};

export function getBrandingLanguageKey(language) {
  return language?.toLowerCase().startsWith('zh') ? 'zh_cn' : 'en';
}

export function resolveLocalizedText(value, language, fallback = '') {
  if (!value || typeof value !== 'object') return fallback;
  const languageKey = getBrandingLanguageKey(language);
  return value[languageKey] || value.en || value.zh_cn || fallback;
}

export function mergeBranding(branding = {}) {
  return {
    ...DEFAULT_BRANDING,
    ...branding,
    site_name: { ...DEFAULT_BRANDING.site_name, ...(branding.site_name || {}) },
    admin_name: { ...DEFAULT_BRANDING.admin_name, ...(branding.admin_name || {}) },
    browser_title: { ...DEFAULT_BRANDING.browser_title, ...(branding.browser_title || {}) },
    browser_description: { ...DEFAULT_BRANDING.browser_description, ...(branding.browser_description || {}) },
    footer_copyright: { ...DEFAULT_BRANDING.footer_copyright, ...(branding.footer_copyright || {}) },
    footer_icp_number: branding.footer_icp_number || DEFAULT_BRANDING.footer_icp_number,
    footer_icp_link: branding.footer_icp_link || DEFAULT_BRANDING.footer_icp_link,
    footer_contact_email: branding.footer_contact_email || DEFAULT_BRANDING.footer_contact_email,
    footer_github_url: branding.footer_github_url || DEFAULT_BRANDING.footer_github_url,
    home_logo_alt: { ...DEFAULT_BRANDING.home_logo_alt, ...(branding.home_logo_alt || {}) },
    home: {
      ...DEFAULT_BRANDING.home,
      ...(branding.home || {}),
      hero: {
        ...DEFAULT_BRANDING.home.hero,
        ...(branding.home?.hero || {}),
        title_prefix: {
          ...DEFAULT_BRANDING.home.hero.title_prefix,
          ...(branding.home?.hero?.title_prefix || {}),
        },
        title_highlight: {
          ...DEFAULT_BRANDING.home.hero.title_highlight,
          ...(branding.home?.hero?.title_highlight || {}),
        },
        title_suffix: {
          ...DEFAULT_BRANDING.home.hero.title_suffix,
          ...(branding.home?.hero?.title_suffix || {}),
        },
        subtitle: {
          ...DEFAULT_BRANDING.home.hero.subtitle,
          ...(branding.home?.hero?.subtitle || {}),
        },
        primary_action: {
          ...DEFAULT_BRANDING.home.hero.primary_action,
          ...(branding.home?.hero?.primary_action || {}),
        },
        secondary_action: {
          ...DEFAULT_BRANDING.home.hero.secondary_action,
          ...(branding.home?.hero?.secondary_action || {}),
        },
      },
      challenge_types: {
        ...DEFAULT_BRANDING.home.challenge_types,
        ...(branding.home?.challenge_types || {}),
        title_prefix: {
          ...DEFAULT_BRANDING.home.challenge_types.title_prefix,
          ...(branding.home?.challenge_types?.title_prefix || {}),
        },
        title_highlight: {
          ...DEFAULT_BRANDING.home.challenge_types.title_highlight,
          ...(branding.home?.challenge_types?.title_highlight || {}),
        },
        subtitle: {
          ...DEFAULT_BRANDING.home.challenge_types.subtitle,
          ...(branding.home?.challenge_types?.subtitle || {}),
        },
      },
      upcoming: {
        ...DEFAULT_BRANDING.home.upcoming,
        ...(branding.home?.upcoming || {}),
        title_prefix: {
          ...DEFAULT_BRANDING.home.upcoming.title_prefix,
          ...(branding.home?.upcoming?.title_prefix || {}),
        },
        title_highlight: {
          ...DEFAULT_BRANDING.home.upcoming.title_highlight,
          ...(branding.home?.upcoming?.title_highlight || {}),
        },
        subtitle: {
          ...DEFAULT_BRANDING.home.upcoming.subtitle,
          ...(branding.home?.upcoming?.subtitle || {}),
        },
        action: {
          ...DEFAULT_BRANDING.home.upcoming.action,
          ...(branding.home?.upcoming?.action || {}),
        },
      },
      leaderboard: {
        ...DEFAULT_BRANDING.home.leaderboard,
        ...(branding.home?.leaderboard || {}),
        title_prefix: {
          ...DEFAULT_BRANDING.home.leaderboard.title_prefix,
          ...(branding.home?.leaderboard?.title_prefix || {}),
        },
        title_highlight: {
          ...DEFAULT_BRANDING.home.leaderboard.title_highlight,
          ...(branding.home?.leaderboard?.title_highlight || {}),
        },
        subtitle: {
          ...DEFAULT_BRANDING.home.leaderboard.subtitle,
          ...(branding.home?.leaderboard?.subtitle || {}),
        },
        action: {
          ...DEFAULT_BRANDING.home.leaderboard.action,
          ...(branding.home?.leaderboard?.action || {}),
        },
      },
    },
  };
}

File: internal/model/branding.go
package model

import (
	"CBCTF/internal/config"
	"database/sql/driver"
	"encoding/json"
	"fmt"
)

const DefaultBrandingCode = "default"

type LocalizedText struct {
	ZhCN string `json:"zh_cn"`
	En   string `json:"en"`
}

func (l LocalizedText) Value() (driver.Value, error) {
	return json.Marshal(l)
}

func (l *LocalizedText) Scan(value any) error {
	if err := scanJSON(value, l); err != nil {
		return fmt.Errorf("failed to scan LocalizedText value")
	}
	return nil
}

type BrandingHeroContent struct {
	TitlePrefix     LocalizedText `json:"title_prefix"`
	TitleHighlight  LocalizedText `json:"title_highlight"`
	TitleSuffix     LocalizedText `json:"title_suffix"`
	Subtitle        LocalizedText `json:"subtitle"`
	PrimaryAction   LocalizedText `json:"primary_action"`
	SecondaryAction LocalizedText `json:"secondary_action"`
}

func (b BrandingHeroContent) Value() (driver.Value, error) {
	return json.Marshal(b)
}

func (b *BrandingHeroContent) Scan(value any) error {
	if err := scanJSON(value, b); err != nil {
		return fmt.Errorf("failed to scan BrandingHeroContent value")
	}
	return nil
}

type BrandingSectionContent struct {
	TitlePrefix    LocalizedText `json:"title_prefix"`
	TitleHighlight LocalizedText `json:"title_highlight"`
	Subtitle       LocalizedText `json:"subtitle"`
}

func (b BrandingSectionContent) Value() (driver.Value, error) {
	return json.Marshal(b)
}

func (b *BrandingSectionContent) Scan(value any) error {
	if err := scanJSON(value, b); err != nil {
		return fmt.Errorf("failed to scan BrandingSectionContent value")
	}
	return nil
}

type BrandingActionSectionContent struct {
	TitlePrefix    LocalizedText `json:"title_prefix"`
	TitleHighlight LocalizedText `json:"title_highlight"`
	Subtitle       LocalizedText `json:"subtitle"`
	Action         LocalizedText `json:"action"`
}

func (b BrandingActionSectionContent) Value() (driver.Value, error) {
	return json.Marshal(b)
}

func (b *BrandingActionSectionContent) Scan(value any) error {
	if err := scanJSON(value, b); err != nil {
		return fmt.Errorf("failed to scan BrandingActionSectionContent value")
	}
	return nil
}

type BrandingHomeContent struct {
	Hero           BrandingHeroContent          `json:"hero"`
	ChallengeTypes BrandingSectionContent       `json:"challenge_types"`
	Upcoming       BrandingActionSectionContent `json:"upcoming"`
	Leaderboard    BrandingActionSectionContent `json:"leaderboard"`
}

func (b BrandingHomeContent) Value() (driver.Value, error) {
	return json.Marshal(b)
}

func (b *BrandingHomeContent) Scan(value any) error {
	if err := scanJSON(value, b); err != nil {
		return fmt.Errorf("failed to scan BrandingHomeContent value")
	}
	return nil
}

type Branding struct {
	Code               string              `gorm:"type:varchar(64);uniqueIndex:idx_brandings_code_active,where:deleted_at IS NULL;not null" json:"code"`
	SiteName           LocalizedText       `gorm:"type:jsonb;not null" json:"site_name"`
	AdminName          LocalizedText       `gorm:"type:jsonb;not null" json:"admin_name"`
	BrowserTitle       LocalizedText       `gorm:"type:jsonb;not null" json:"browser_title"`
	BrowserDescription LocalizedText       `gorm:"type:jsonb;not null" json:"browser_description"`
	FooterCopyright    LocalizedText       `gorm:"type:jsonb;not null" json:"footer_copyright"`
	HomeLogo           FileURL             `json:"home_logo"`
	HomeLogoAlt        LocalizedText       `gorm:"type:jsonb;not null" json:"home_logo_alt"`
	Home               BrandingHomeContent `gorm:"type:jsonb;not null" json:"home"`
	BaseModel
}

func DefaultBranding() Branding {
	return Branding{
		Code: DefaultBrandingCode,
		SiteName: LocalizedText{
			ZhCN: "深潜 CTF",
			En:   "DEEP DIVE CTF",
		},
		AdminName: LocalizedText{
			ZhCN: "深潜管理台",
			En:   "DEEP DIVE Admin",
		},
		BrowserTitle: LocalizedText{
			ZhCN: "深潜 CTF",
			En:   "DEEP DIVE CTF",
		},
		BrowserDescription: LocalizedText{
			ZhCN: "深潜 CTF 网络安全竞赛平台",
			En:   "DEEP DIVE CTF competition platform",
		},
		FooterCopyright: LocalizedText{
			ZhCN: "© 2025 深潜 CTF",
			En:   "© 2025 DEEP DIVE CTF",
		},
		HomeLogo: FileURL(fmt.Sprintf("%s/platform/logo.png", config.Env.Host)),
		HomeLogoAlt: LocalizedText{
			ZhCN: "深潜 CTF 首页 Logo",
			En:   "DEEP DIVE CTF home logo",
		},
		Home: BrandingHomeContent{
			Hero: BrandingHeroContent{
				TitlePrefix: LocalizedText{
					ZhCN: "深入探索",
					En:   "Dive Deep into the",
				},
				TitleHighlight: LocalizedText{
					ZhCN: "网络安全",
					En:   "Cyber Security",
				},
				TitleSuffix: LocalizedText{
					ZhCN: "挑战",
					En:   "Challenge",
				},
				Subtitle: LocalizedText{
					ZhCN: "加入深潜 CTF 社区, 在真实场景中对抗、练习与成长",
					En:   "Join the elite community of hackers, compete in real-world challenges, and master the art of cybersecurity through hands-on experience.",
				},
				PrimaryAction: LocalizedText{
					ZhCN: "立即参赛",
					En:   "START HACKING",
				},
				SecondaryAction: LocalizedText{
					ZhCN: "了解更多",
					En:   "LEARN MORE",
				},
			},
			ChallengeTypes: BrandingSectionContent{
				TitlePrefix: LocalizedText{
					ZhCN: "掌握",
					En:   "Master All Aspects of",
				},
				TitleHighlight: LocalizedText{
					ZhCN: "网络安全",
					En:   "Cyber Security",
				},
				Subtitle: LocalizedText{
					ZhCN: "从 Web 渗透到二进制分析, 覆盖网络安全核心方向",
					En:   "From web exploitation to binary analysis, our challenges cover every major domain of cybersecurity",
				},
			},
			Upcoming: BrandingActionSectionContent{
				TitlePrefix: LocalizedText{
					ZhCN: "近期",
					En:   "Upcoming",
				},
				TitleHighlight: LocalizedText{
					ZhCN: "赛事",
					En:   "Competitions",
				},
				Subtitle: LocalizedText{
					ZhCN: "报名即将开启的 CTF 赛事, 与高手同场竞技",
					En:   "Register now for our upcoming CTF events and compete with the best",
				},
				Action: LocalizedText{
					ZhCN: "查看全部赛事",
					En:   "VIEW ALL CONTESTS",
				},
			},
			Leaderboard: BrandingActionSectionContent{
				TitlePrefix: LocalizedText{
					ZhCN: "顶尖",
					En:   "Top",
				},
				TitleHighlight: LocalizedText{
					ZhCN: "战队",
					En:   "Performers",
				},
				Subtitle: LocalizedText{
					ZhCN: "查看当前积分榜上的领先队伍",
					En:   "Meet the elite teams leading our global leaderboard",
				},
				Action: LocalizedText{
					ZhCN: "查看排行榜",
					En:   "VIEW SCOREBOARD",
				},
			},
		},
	}
}
